### PR TITLE
Add Rules logging and seed data PUT request format fix 

### PR DESCRIPTION
**Pull request by jillcary**
# Type of change? 

- [x] Bug fix
- [ ] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 
Seed data was unable to create rules for the java microservice.
The java microservice was expecting a "dateCreated" value in the PUT request and was not receiving this value on seed data creation.

Also see related issue for future refactoring of the DateCreated field:
https://github.com/Azure/device-telemetry-java/issues/103

**Checklist:**
- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [x] The change requires a change to the documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 1]
```diff
@@ -175,28 +201,53 @@ private RuleServiceModel requestBodyToRuleServiceModel(JsonNode body, String id)
                     ConditionApiModel.class).toServiceModel());
             }
         } catch (Exception e) {
+            log.error("Invalid input, " +
```
Comment (zhang-hua) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: suggest to refactor this line to reduce duplicate code.
String message = String.format("...")

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 2]
```diff
@@ -175,28 +201,53 @@ private RuleServiceModel requestBodyToRuleServiceModel(JsonNode body, String id)
...
 
         // update existing Rule if ETag is present
-        if (body.has("ETag") || body.has("etag")) {
+        if (body.has("ETag") ||
+            body.has("etag") ||
+            body.has("Etag")) {
+
+            // Rule ID must be specified in the request e.g. rules/{put-rule-id}
+            if (id == null) {
+                log.error("Bad request -- null rule id in PUT request " +
```
Comment (zhang-hua) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: suggest to refactor this line to reduce duplicate code.
String message = String.format("...")

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 3]
```diff
@@ -290,8 +291,12 @@ public Rules(
                 }
 
                 try {
-                    return getServiceModelFromJson(
-                        Json.parse(result.getBody()));
+                    RuleServiceModel rule =
+                        getServiceModelFromJson(Json.parse(result.getBody()));
+
+                    log.info("Sucessfully retrieved rule id " + rule.getId());
```
Comment (xzsheng) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: Typo: successfully

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 4]
```diff
@@ -206,9 +257,13 @@ private RuleServiceModel requestBodyToRuleServiceModel(JsonNode body, String id)
                     JsonHelper.getNode(body, "Severity").asText(),
                     conditions);
             } catch (Exception e) {
-                throw new Exception("Invalid input");
+                log.error("Invalid input for new rule with id. Error msg: " + e.getMessage(), e);
+                throw new Exception("Invalid input. Error msg: " + e.getMessage(), e);
```
Comment (zhang-hua) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: use 'InvalidInputException' to throw this?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 5]
```diff
@@ -218,7 +273,8 @@ private RuleServiceModel requestBodyToRuleServiceModel(JsonNode body, String id)
                     JsonHelper.getNode(body, "Severity").asText(),
                     conditions);
             } catch (Exception e) {
-                throw new Exception("Invalid input");
+                log.error("Invalid input for new rule. Error msg: " + e.getMessage(), e);
+                throw new Exception("Invalid input. Error msg: " + e.getMessage(), e);
```
Comment (zhang-hua) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: use 'InvalidInputException' to throw this?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 6]
```diff
@@ -175,28 +201,53 @@ private RuleServiceModel requestBodyToRuleServiceModel(JsonNode body, String id)
...
+                    DateTime.now(DateTimeZone.UTC).toString(DATE_FORMAT),
                     JsonHelper.getNode(body, "Enabled").asBoolean(),
                     JsonHelper.getNode(body, "Description").asText(),
                     JsonHelper.getNode(body, "GroupId").asText(),
                     JsonHelper.getNode(body, "Severity").asText(),
                     conditions);
             } catch (Exception e) {
-                throw new Exception("Invalid input");
+                log.error("Invalid input for rule with ETag. Error msg: " + e.getMessage(), e);
+                throw new Exception("Invalid input. Error msg: " + e.getMessage(), e);
```
Comment (zhang-hua) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: use 'InvalidInputException' to throw this?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 7]
```diff
@@ -175,28 +201,53 @@ private RuleServiceModel requestBodyToRuleServiceModel(JsonNode body, String id)
                     ConditionApiModel.class).toServiceModel());
             }
         } catch (Exception e) {
+            log.error("Invalid input, " +
+                "request does not contain the `Conditions` field. Error msg: " +
+                e.getMessage(), e);
             throw new Exception("Invalid input, " +
```
Comment (zhang-hua) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: use 'InvalidInputException' to throw this?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 8]
```diff
@@ -175,28 +201,53 @@ private RuleServiceModel requestBodyToRuleServiceModel(JsonNode body, String id)
...
-        if (body.has("ETag") || body.has("etag")) {
+        if (body.has("ETag") ||
+            body.has("etag") ||
+            body.has("Etag")) {
+
+            // Rule ID must be specified in the request e.g. rules/{put-rule-id}
+            if (id == null) {
+                log.error("Bad request -- null rule id in PUT request " +
+                    "with etag in body. Must specify rule id in request URL.");
+                throw new Exception("Bad request -- null rule id in PUT request " +
```
Comment (zhang-hua) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: use 'BadRequestException' to throw this?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java [hunk 9]
```diff
@@ -175,28 +201,53 @@ private RuleServiceModel requestBodyToRuleServiceModel(JsonNode body, String id)
...
+                    " from json: " + body);
+
+                // TODO: pull date created from backend for existing rules
+                // https://github.com/Azure/device-telemetry-java/issues/103
                 result = new RuleServiceModel(
                     JsonHelper.getNode(body, "ETag").asText(),
                     id,
                     JsonHelper.getNode(body, "Name").asText(),
-                    JsonHelper.getNode(body, "DateCreated").asText(),
+                    DateTime.now(DateTimeZone.UTC).toString(DATE_FORMAT),
```
Comment (xzsheng) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: How about set the default value in constructor, just like what we did for `dateModified` ?

**Comment by xzsheng** [comment]
It looks we are going to add log on enter, exit and exception for lots of methods. It may hurt the code readability. Is it possible to leverage some AOP technique for this purpose?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java (hunk 4 above)
Comment (jillcary) [reply app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: The controllers for our java implementation only have BadRequestExceptions, I think invalid input is in the service side exceptions. I've updated to use BadRequestException, does that work?

**Comment by jillcary** [comment]
Created a new issue for investigating AOP techniques, this is something we should use across all Java microservices eventually if possible: 
https://github.com/Azure/device-telemetry-java/issues/106

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java (hunk 1 above)
Comment (jillcary) [reply app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/RulesController.java]: Agreed, fixing throughout.

**Comment by dluc** [comment]
wrt AOP [this](http://autofaccn.readthedocs.io/en/latest/advanced/interceptors.html) is one way of doing it, just be careful with the perf impact

**Pull request merged**